### ISSUE TITLE
Project: Makeflow JSON Representation

### ISSUE THREAD

**Issue by dthain** [cooperative-computing-lab/cctools#1178]
Modify makeflow to emit the DAG in a straightforward JSON representation.  This will be the starting point for some interesting modifications and enhancements to Makeflow.

Start by modifying `makeflow/src/dag_visitor.c` and add a new visitor dag_to_json that takes a rule like this:

```
CPUS=5
MEMORY=10
DISK=20
output: input calib.dat
    mysim.exe input > output
```

And outputs it something like this: (details can be changed)

```
{
   cmd: "mysim.exe input >output",
   inputs: [ "input", "calib.dat" ],
   outputs: [ "output" ],
   environment: { "PATH": "/usr/bin" ],
   cores 5,
   memory: 10,
   disk: 20
}
```

Important: use the JX library in `dttools` to generate jx objects and then write them out to a file.  This will ensure that everything is escaped and constructed correctly.

I _think_ the overall structure of the file should be a JSON array.  Or perhaps it should be a dictionary, with the job id as the key?  I'm not sure.  Try one out, and then let's see how things look.

**Comment by dthain** [comment]
@trshaffer take a crack at this one.

**Comment by dthain** [comment]
Ok, we now have a first draft of exporting a Makelfow as JSON.
Next step is to read back that JSON file and convert it to a makeflow dag internally.
(Will require writing out a spec of exactly what we accept.)

**Comment by trshaffer** [comment]
I can write out a description of the output format I've been using, which ought to lead into a discussion on parsing

**Comment by trshaffer** [comment]
If I were parsing the JSON output back into a DAG, this is roughly what I'd require. A lot of the decisions (e.g. layout, which keys are required) are based on the internal data structures or current parser behavior. The variables/environment are a bit of a rough patch. I made some comments in italics in the relevant sections

# JSON Representation of Makeflows #

A Makeflow is represented as a JSON object with the following required keys.

- `"categories"`
- `"default_category"`
- `"rules"`

In addition, the following key is optional.

- `"environment"`

`"default_category"` must be one of the string keys of the categories object. The other values are described in the following sections.

## Categories ##

The categories of a Makeflow are represented as a JSON object mapping category names to category objects. Each category object contains the following keys indicating the resources required by the category.
*When the Makeflow is parsed, these are filled in with the host resources, so these are always available. Should these be optional?*

- `"cores"`
- `"disk"`
- `"memory"`
- `"gpus"`

A category object may also contain a `"variables"` key, whose value is an object mapping variable names to string values. These name -> value mappings are available as variables in rules under this category.

## Environment ##

The environment object in a Makeflow represents a set of variables that should be exported to the environment when running jobs. The values of this object are currently `null`, but might be used to store environment variables from the host in a future revision.
*These are stored as a set in `struct dag`. When parsing, Makeflow stores the host environment as variables in the default category. This arrangement feels a little awkward to me. It might be nice if each category (and possibly each rule) could have its own distinct substitution variables and export variables.*

## Rules ##

The production rules in a Makeflow are stored as an array containing one or more rule objects. Note that the order of rules within the array is arbitrary; Makeflow reads in all the rules and determines a correct order of execution. Each rule object is required to have the following keys.

- `"category"`: a string indicating the category of this rule. Must match one of the keys in the category object of the Makeflow.
- `"inputs"`: list of files needed by the rule
- `"outputs"`: list of files created by the rule
- `"allocation"`: category allocation used by the rule. Must be one of `"unlabeled"`, `"user"`, `"auto_zero"`, `"auto_first"`, `"auto_max"`, or `"error"` *(?)*

File lists are represented as JSON arrays containing zero or more strings.
*Turns out the parser accepts rules with no inputs or outputs.*

The following keys are optional.

- `"local_job"`: a boolean indicating whether or not the rule must be run locally. Defaults to `false`
- `"resources_needed"`: a JSON object indicating the resource requirements for this rule. Uses the same `"cores"`, `"disk"`, `"memory"`, `"gpus"` as category objects.
- `"remote_names"`: a JSON object containing local name -> remote name mappings as strings.

Each rule object must also contain exactly one of the following keys.

- `"command"`: a string command to run
- `"makeflow"`: a JSON object containing the following keys. Indicates the sub-Makeflow to run in this rule.
  *Should we accept an actual JSON object as a parsed sub-Makeflow?*
  + `"path"`: a string indicating the location of the sub-Makeflow file
  + `"cwd"`: the working directory in which to run the sub-Makeflow

**Comment by dthain** [comment]
This is a good start.  Some minor suggestions:

- Instead of requiring `"default_category"`, how about require a category with the name `default`?
- How do we represent tasks that have a different name at the DAG level and at the task level?
- How is this similar/different from the job specification in Confuga?

@btovar any thoughts?

**Comment by btovar** [comment]
work queue and makeflow have that "default" category if the task/rule is not explicitly labeled with one.

**Comment by batrick** [comment]
> Remote renaming appears to map cleanly here. We might want to allow the command to be an array of arguments as in Chirp job syntax. Internally, Makeflow just stores a string, but I could see wrappers operating on arguments.

Makeflow (and Make) recipes are just shell code (possibly more than one command!). I think it would be wise to not interpret the string in any way unless there is a very good reason.

Another thing to consider related to remote renaming but maybe outside the scope of this issue: Makeflow runs with a `--working-dir` option that sets the "current directory" of Makeflow relative to the Chirp server root. So if a rule has a dependency `a`, its actual path (written in the Chirp job description) might be `/home/patrick/workflow1/a`. So, remote renaming (if added to Chirp) needs to take this working directory into account somehow. [Also, an unfortunate problem Makeflow+Chirp has right now is that the working directory is prepended to **all** job files which makes it impossible to refer to files outside the working directory on the Chirp server without hacks (like symbolic links). With the benefit of hindsight, it would have been better to do that only for files with a relative path.]

**Comment by dthain** [comment]
I like how you have introduced variable substitutions, which will be a big improvement on the macro substitutions that we have now. But we still need to be able to define environment variables for individual tasks as well as the workflow.  What does that look like?

**Comment by trshaffer** [comment]
At the moment, it looks like export variables are defined on the DAG itself. I reflected that in the original format with the awkward environment set/map. When the rules run, the environment gets created by looking up each global export variable in the node, category, DAG, and finally host environment. So at the moment, it looks like Makeflow only has substitution variables and doesn't support per-rule or per-category export variables, only locally changing the values of the global set of export variables.

I'd like to see each category and rule get its own map of environment variables, but I think this would require some internal changes to Makeflow.

**Comment by btovar** [comment]
Tim, that is correct. It is a little awkward. Variables can be set per rule or category, but whether they are exported is defined per dag.

**Comment by trshaffer** [comment]
As a workaround while parsing the JSON representation, I'm inserting export variables on their corresponding rules/categories and adding all the names to the DAG-wide export set. Is this the best approach @btovar, or should we be looking at changing the internal representation. I think this workaround *should* be fine in all but pathological cases, and changing the variable handling looks like it would be a little invasive

**Comment by btovar** [comment]
I like the that variables are added per rule, even on repetitions. I am not sure about the meaning of dag-wide environment variables, though. (I know this is how makeflow does it internally, but perhaps we can start cleaning that up with the JSON representation?)

**Comment by trshaffer** [comment]
I'm not using Makeflow's internal variables for much besides export variables and resource stuff at the moment, so I don't think it should be possible to specify a non-export Makeflow variable and get unexpected behavior. Substitution variables will be handled during parse by evaluating JX expressions, so I can *probably* avoid changing internals. This being more of an issue in parsing the Make-style syntax, do you think cleaning up variable handling would be worth it @btovar @dthain ?

**Comment by dthain** [comment]
You guys know the code better than me at this point, so I will leave it up to you.

I'll just suggest that simplicity is probably more important than efficiency at this point.

**Comment by dthain** [comment]
Implemented in PR #1683
